### PR TITLE
Fix flow static typing

### PR DESCRIPTION
**Pull request by TomaszKolek**

### CODE REVIEW

File: src/api/apiFetch.js [hunk 1]
```diff
@@ -27,7 +28,7 @@ export const apiCall = async (
   auth: Auth,
   route: string,
   params: Object = {},
-  resFunc = res => res,
+  resFunc: (Json) => mixed = res => res,
```
Comment (TomaszKolek) [inline src/api/apiFetch.js]: is `Json` a good name?

File: src/api/apiFetch.js [hunk 2]
```diff
@@ -60,31 +61,56 @@ export const apiCall = async (
...
+export const apiGet = async (
+  auth: Auth,
+  route: string,
+  params: Object = {},
+  resFunc: (Json) => mixed,
+) =>
   apiCall(auth, `${route}?${encodeAsURI(params)}`, {
     method: 'get',
-  }, resFunc, options);
+  }, resFunc);
```
Comment (TomaszKolek) [inline src/api/apiFetch.js]: I removed options because it looked like a bug. There is no such thing also I haven't seen any usage of this paramter

File: src/api/emojiReactionAdd.js [hunk 3]
```diff
@@ -9,5 +10,6 @@ export default (
   apiPut(
     auth,
     `messages/${messageId}/emoji_reactions/${emojiName}`,
+    {},
```
Comment (TomaszKolek) [inline src/api/emojiReactionAdd.js]: I added empty object here because it looked that function was passing as a parameter object.

File: src/api/apiFetch.js (hunk 1 above)
Comment (borisyankov) [inline src/api/apiFetch.js]: Nope it isn't reall :)
Also not sure this is even correct way to declare this. My guess is not.

Comment (TomaszKolek) [inline src/api/apiFetch.js]: Do you mean from `flow` point of view? I think it is.

File: src/api/apiFetch.js (hunk 2 above)
Comment (borisyankov) [inline src/api/apiFetch.js]: Yeah, that really looks suspicious. Looks like the first bug found by you and Flow :)

File: src/api/emojiReactionAdd.js (hunk 3 above)
Comment (borisyankov) [inline src/api/emojiReactionAdd.js]: Not ideal. You should mark the apiPut parameter as optional.

Comment (borisyankov) [inline src/api/emojiReactionAdd.js]: `hello: number' vs `hello?: number`

Comment (TomaszKolek) [inline src/api/emojiReactionAdd.js]: but its not a declaration its execution - is there any way in JS to pass a parameter with the name like:

```
getSomething("parameter1", paramter3="parameter3")
```

Comment (borisyankov) [inline src/api/emojiReactionAdd.js]: I mean you can set that in the `apiPut` declaration itself.

File: .flowconfig [hunk 4]
```diff
@@ -49,5 +48,10 @@ suppress_comment=\\(.\\|\n\\)*\\$FlowExpectedError
 
 unsafe.enable_getters_and_setters=true
 
+module.file_ext=.ios.js
+module.file_ext=.js
+module.file_ext=.jsx
```
Comment (borisyankov) [inline .flowconfig]: Do we have the jsx files somewhere?

File: .eslintrc [hunk 5]
```diff
@@ -44,6 +44,7 @@
     "react/prefer-stateless-function": 0,
     "react/no-array-index-key": 0,
     "react/no-unused-prop-types": 0,
+    "react/require-default-props": 0,
```
Comment (borisyankov) [inline .eslintrc]: I remember there was a reason for that, but now I am thinking it is also useful.
It will be ideal, if we can avoid ignoring this rule.

File: package.json [hunk 6]
```diff
@@ -67,7 +67,7 @@
     "eslint-plugin-react": "~7.0.0",
     "eslint-plugin-react-native": "^2.2.1",
     "eslint-plugin-spellcheck": "~0.0.6",
-    "flow-bin": "~0.40.0",
+    "flow-bin": "^0.40.0",
```
Comment (borisyankov) [inline package.json]: I think we have to have a strict '0.40' so the `~` was intentional. That is what RN is on.
Did you need the newer ones?

File: src/api/apiFetch.js [hunk 7]
```diff
@@ -60,31 +61,56 @@ export const apiCall = async (
...
 
-export const apiPost = async (auth, route, params = {}, resFunc, options) =>
+export const apiPost = async (
+  auth: Auth,
+  route: string,
+  resFunc: (ApiResponse) => mixed,
```
Comment (borisyankov) [inline src/api/apiFetch.js]: You can define a new type ResponseExtractor (ResponseExtractionFunc?) and reuse it.

File: src/api/apiFetch.js [hunk 8]
```diff
@@ -60,31 +61,56 @@ export const apiCall = async (
...
     method: 'post',
     body: encodeAsURI(params),
-  }, resFunc, options);
+  }, resFunc);
 
-export const apiPut = async (auth, route, params = {}, resFunc, options) =>
+export const apiPut = async (
+  auth: Auth,
+  route: string,
+  resFunc: (ApiResponse) => mixed,
```
Comment (borisyankov) [inline src/api/apiFetch.js]: And the type from before will be reused here.

File: src/api/devFetchApiKey.js [hunk 9]
```diff
@@ -1,9 +1,11 @@
...
+import { Auth } from '../types';
+import { apiPost } from './apiFetch';
 
 export default (auth: Auth, email: string) =>
   apiPost(
     auth,
     'dev_fetch_api_key',
-    { username: email },
     res => res.api_key,
+    { username: email },
```
Comment (borisyankov) [inline src/api/devFetchApiKey.js]: Was that in the wrong position?

File: src/api/messagesFlags.js [hunk 10]
```diff
@@ -1,14 +1,16 @@
...
+import { Auth } from '../types';
+import { apiPost } from './apiFetch';
 
 export default (
   auth: Auth,
   messages: number[],
   op: string,
   flag: string,
-): number[] =>
+): mixed =>
```
Comment (borisyankov) [inline src/api/messagesFlags.js]: If it is possible to declare the response more concretely than `mixed` that would be important to do.

File: src/types.js [hunk 11]
```diff
@@ -85,3 +85,8 @@ export type Recipient = {
   subject: string,
   email: string,
 };
+
+export type ApiResponse = {
```
Comment (borisyankov) [inline src/types.js]: We would want one type per API call so we can get better catch issues.

File: .flowconfig (hunk 4 above)
Comment (TomaszKolek) [inline .flowconfig]: Not sure - .ios.js seems to be the most important.

File: src/api/devFetchApiKey.js (hunk 9 above)
Comment (TomaszKolek) [inline src/api/devFetchApiKey.js]: we changed order of parameters

File: .eslintrc (hunk 5 above)
Comment (TomaszKolek) [inline .eslintrc]: I don't see any other way of doing it - as I mentioned yesterday, people suggest that we can ignore it here because flow is actualy taking care about it.

File: package.json (hunk 6 above)
Comment (TomaszKolek) [inline package.json]: Done

File: src/api/apiFetch.js (hunk 7 above)
Comment (TomaszKolek) [inline src/api/apiFetch.js]: good idea!

File: package.json (hunk 6 above)
Comment (neerajwahi) [inline package.json]: Looks like it's still matching any minor version greater than 0.40? Let's leave the `~` for now and change it in another diff

File: src/account-info/AccountDetailsScreen.js [hunk 12]
```diff
@@ -1,38 +1,57 @@
...
+  fullName: string;
+  avatarUrl: string;
+  status: string;
 
   componentWillMount() {
     // props.email gets reset during navigation slide out (on back)
-    // so we cache the value to prevent an exception
+    // so we cache values to prevent an exception
     this.email = this.props.email;
+    const user = this.props.users.find(x => x.email === this.email) || NOT_FOUND_USER;
```
Comment (neerajwahi) [inline src/account-info/AccountDetailsScreen.js]: I would refactor this to just cache `this.user` and then all the fields you want (such as `email`, `fullName`, etc..) you can pull from inside of render. Does my explanation make sense?

File: src/api/messagesFlags.js (hunk 10 above)
Comment (TomaszKolek) [inline src/api/messagesFlags.js]: It will be `any` not mixed. Because our resFunc can return `any` here we alse have to have the same type.

File: src/api/apiFetch.js (hunk 2 above)
Comment (neerajwahi) [inline src/api/apiFetch.js]: 👍

File: src/api/messagesFlags.js (hunk 10 above)
Comment (neerajwahi) [inline src/api/messagesFlags.js]: 👍 to this

We can also leave it like this for now and add a TODO to come back to it if it's tricky

File: src/types.js (hunk 11 above)
Comment (TomaszKolek) [inline src/types.js]: I'd keep it for now - it'd just more code with any additional reason because responses for now looks almost the same.

File: src/api/uploadFile.js [hunk 13]
```diff
@@ -1,14 +1,15 @@
...
+import { Auth } from '../types';
+import { apiPost } from './apiFetch';
 
 export default (
   auth: Auth,
   messages: number[],
   op: string,
   flag: string,
-): number[] =>
+): mixed =>
```
Comment (neerajwahi) [inline src/api/uploadFile.js]: Can you add a TODO to remove the `mixed` return type here?

File: src/common/ImageAvatar.js [hunk 14]
```diff
@@ -11,7 +12,22 @@ const styles = StyleSheet.create({
...
+  avatarUrl: string,
+  size: number,
+  status?: string,
+  isCircular?: boolean,
+  onPress?: () => void
+}
+
+
+export default({
+                 avatarUrl,
```
Comment (neerajwahi) [inline src/common/ImageAvatar.js]: Can we adjust the format of this block to something like:
```
export default({
  avatarUrl,
  size,
  ...,
}: funcArguments) => {
  ...
}
```
It's just that it's more consistent with the rest of the code :)

File: src/html/HtmlNodeTag.js [hunk 15]
```diff
@@ -42,7 +44,16 @@ const specialTags = {
 const stylesFromClassNames = (classNames = '', styleObj) =>
   classNames.split(' ').map(className => styleObj[className]);
 
-export default ({ auth, attribs, name, cascadingStyle, childrenNodes, onPress }) => {
+type FuncArguments = {
```
Comment (neerajwahi) [inline src/html/HtmlNodeTag.js]: Looks like we have a mix of `funcArguments` and `FuncArguments` in different places. Could we standardize to one of them? I think the convention is `FuncArguements` but could be mistaken

File: src/types.js (hunk 11 above)
Comment (neerajwahi) [inline src/types.js]: We can add a TODO and do this in another PR as well to get things merged quickly :)

File: .flowconfig (hunk 4 above)
Comment (TomaszKolek) [inline .flowconfig]: Done

File: src/html/HtmlNodeTag.js (hunk 15 above)
Comment (TomaszKolek) [inline src/html/HtmlNodeTag.js]: Done

File: src/account-info/AccountDetailsScreen.js (hunk 12 above)
Comment (TomaszKolek) [inline src/account-info/AccountDetailsScreen.js]: Hope I understood it - let you know when it's ready

File: src/account-info/AccountDetailsScreen.js [hunk 16]
```diff
@@ -1,38 +1,50 @@
...
+    popRoute: (string) => void
+  };
+  user: Object;
 
   componentWillMount() {
     // props.email gets reset during navigation slide out (on back)
-    // so we cache the value to prevent an exception
-    this.email = this.props.email;
+    // so we cache values to prevent an exception
+    this.user = this.props.users.find(x => x.email === this.props.email) || NOT_FOUND_USER;
```
Comment (TomaszKolek) [inline src/account-info/AccountDetailsScreen.js]: @borisyankov @neerajwahi I refactored as @neerajwahi suggested - take a look at it please
